include/execution.hpp

```diff
@@ -2120,7 +2221,99 @@ namespace std::execution {
   /////////////////////////////////////////////////////////////////////////////
   // [execution.senders.adaptors.upon_stopped]
   namespace __upon_stopped {
+
+    template <class _ReceiverId, class _Fun>
+      class __receiver
+        : receiver_adaptor<__receiver<_ReceiverId, _Fun>, __t<_ReceiverId>> {
+        using _Receiver = __t<_ReceiverId>;
+        friend receiver_adaptor<__receiver, _Receiver>;
+        [[no_unique_address]] _Fun __f_;
+
+        // Customize set_stopped by invoking the invocable and passing the result
+        // to the base class
+        template <class _R2 = _Receiver>
+          requires invocable<_Fun> &&
+            __callable<set_value_t, _R2, invoke_result_t<_Fun>>
+        void set_stopped() && noexcept try {
+          execution::set_value(
+              ((__receiver&&) *this).base(),
+              std::invoke((_Fun&&) __f_));
+        } catch(...) {
+          execution::set_error(
+              ((__receiver&&) *this).base(),
+              current_exception());
+        }
+        // Handle the case when the invocable returns void
+        template <class _R2 = _Receiver>
+          requires invocable<_Fun> &&
+            same_as<void, invoke_result_t<_Fun>> &&
+              __callable<set_value_t, _R2>
+        void set_stopped() && noexcept try {
+          invoke((_Fun&&) __f_);
+          execution::set_value(((__receiver&&) *this).base());
+        } catch(...) {
+          execution::set_error(
+              ((__receiver&&) *this).base(),
+              current_exception());
+        }
+
+       public:
+        explicit __receiver(_Receiver __rcvr, _Fun __fun)
+          : receiver_adaptor<__receiver, _Receiver>((_Receiver&&) __rcvr)
+          , __f_((_Fun&&) __fun)
+        {}
+      };
+
+    template <class _SenderId, class _Fun>
+      class __sender
+        : sender_adaptor<__sender<_SenderId, _Fun>, __t<_SenderId>> {
+        using _Sender = __t<_SenderId>;
+        friend sender_adaptor<__sender, _Sender>;
+        template <class _Receiver>
+          using __receiver = __receiver<__x<remove_cvref_t<_Receiver>>, _Fun>;
+
+        [[no_unique_address]] _Fun __fun_;
+
+          using __set_stopped =
+            completion_signatures<
+              __minvoke1<
+                __remove<void, __qf<set_value_t>>,
+                invoke_result_t<_Fun>>>;
+
+        template <class _Env>
+          using __completion_signatures =
+            make_completion_signatures<
+              _Sender, _Env, __with_exception_ptr, 
```

Comment:
I was wondering about the requirement on `set_error_t(exception_ptr)` to be valid.

I *think* we can inside the catch bock add an `if constexpr` that checks whether `set_error_t` is `tag_invocable` with whatever exception we got

Reply:
We wouldn't want to do that since it might try to instantiate the error channel. What we want is to avoid instantiating _any_ channel that will be unused, based on the sender's completion signatures.